Add multi-client support in SOF

ranj063 opened the pull request:
This is the first half of the multi-client support series. This series only really creates one client for audio but the changes are quite invasive. 

This will be followed with another series that adds other clients.

RanderWang commented on sound/soc/sof/sof-pci-dev.c:
```diff
@@ -152,6 +157,7 @@ static const struct sof_dev_desc cml_desc = {
 	.chip_info = &cnl_chip_info,
 	.default_fw_path = "intel/sof",
 	.default_tplg_path = "intel/sof-tplg",
+	.default_fw_filename = "sof-cnl.ri",
```
here, you may need to use sof-cml.ri later

ranj063 replied:
@RanderWang sure, I will rebase after the other PR is merged.

dbaluta commented on sound/soc/sof/sof-acpi-dev.c:
```diff
@@ -177,35 +176,9 @@ static int sof_acpi_probe(struct platform_device *pdev)
 	}
 
-#if IS_ENABLED(CONFIG_SND_SOC_SOF_FORCE_NOCODEC_MODE)
-	/* force nocodec mode */
```
@ranj063 ops variable is no longer needed. Please remove it together with associated code.

ranj063 replied:
@dbaluta ack!

dbaluta commented:
@ranj063 I have sent you on email some patches to make the corresponding changes for IMX. I hope you can squash them to this PR.

Then I have sent a first draft of the direction I want to follow to unify DT / ACPI stuff. I hope you can comment on that. 

The idea is that we need to remove access to ACPI specific snd_soc_acpi_mach and replace it with a generic structure.

ranj063 commented:
> @ranj063 I have sent you on email some patches to make the corresponding changes for IMX. I hope you can squash them to this PR.
> 
> Then I have sent a first draft of the direction I want to follow to unify DT / ACPI stuff. I hope you can comment on that.
> 
> The idea is that we need to remove access to ACPI specific snd_soc_acpi_mach and replace it with a generic structure.

@dbaluta thanks! I will incorporate your changes into the PR

ranj063 commented:
> The idea is that we need to remove access to ACPI specific snd_soc_acpi_mach and replace it with a generic structure.

@dbaluta I've added your changes in and added 2 new commits to support the unified machine drv descriptor. Can you please take a look?

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -24,7 +24,7 @@ static int create_page_table(struct snd_soc_component *component,
 {
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_sof_dev *sdev = dev_get_drvdata(component->dev->parent);
```
Sorry, but I have allergic reactions when we don't have a clear isolation between parent and child devices.
we shouldn't do this, this is a symptom that using sdev everywhere was a convenient solution that has exceeded its shelf life. Let's clean this up.

ranj063 replied:
@plbossart sure yes this is a bit ugly. I need to separate out the client (child) from the parent a bit more. will fix in the next version

plbossart replied:
we should really separate out structures. Maybe first split the structure, even before the introduction of the clients?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,139 @@
+#else
+	/* fallback to nocodec mode */
+	dev_warn(sdev->dev, "No ASoC machine driver found - using nocodec\n");
+	machine = devm_kzalloc(sdev->dev, sizeof(*machine), GFP_KERNEL);
```
again, this looks like we are allocating dev-managed memory in the scope of the parent (e.g PCI device) when this is solely used by the child.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,139 @@
+static int sof_audio_probe(struct platform_device *pdev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(pdev->dev.parent);
+	struct snd_sof_pdata *plat_data = sdev->pdata;
```
no, let's not do this.
We can pass data to a platform-device. let's not directly modify the memory handled by the parent.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,139 @@
+	pm_runtime_use_autosuspend(&pdev->dev);
+	pm_runtime_enable(&pdev->dev);
+	pm_runtime_mark_last_busy(&pdev->dev);
+	pm_runtime_put_noidle(&pdev->dev);
```
don't you have to allow pm_runtime for a platform device?

ranj063 replied:
@plbossart not for platform devices. It is only needed for PCI devices.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,139 @@
+	pm_runtime_use_autosuspend(&pdev->dev);
+	pm_runtime_enable(&pdev->dev);
+	pm_runtime_mark_last_busy(&pdev->dev);
+	pm_runtime_put_noidle(&pdev->dev);
```
also why put_noidle here? we use put_autosuspend usually

ranj063 replied:
fixed

dbaluta commented:
> > The idea is that we need to remove access to ACPI specific snd_soc_acpi_mach and replace it with a generic structure.
> 
> @dbaluta I've added your changes in and added 2 new commits to support the unified machine drv descriptor. Can you please take a look?

@ranj063 I like the way you figured out how to fill the type. It's clever. I was thinking to add another field inside desc structure.

Let me give it a test tomorrow and fill in the TODO spaces. Also, after @plbossart comments I think would be a good idea to have a skype meeting next week.

ranj063 commented:
> sorry @ranj063 I don't get the parent/child definitions and the widespread use of component->dev->parent.

@plbossart I have addressed this with the last 2 commits in the series. Unfortunately, there's no easy way for me to remove the references to the parent from the client device right from the get go. But if you get to the end of the series, I have gradually removed all references and eventually made sof-priv.h exclusive for the core and platform-specific code.

ranj063 commented:
> Let me give it a test tomorrow and fill in the TODO spaces. Also, after @plbossart comments I think would be a good idea to have a skype meeting next week.

@dbaluta sure, thanks a bunch. I have addressed @plbossart concerns in the series as well. I might need to do re-ordering/squashing later on but this way it will be easy to review.

ranj063 commented:
SOFCI TEST

ranj063 commented:
SOFCI TEST

kv2019i commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -88,7 +88,7 @@ int hda_dsp_pcm_hw_params(struct snd_soc_component *scomp,
 			  struct sof_ipc_stream_params *ipc_params)
 {
-	struct snd_sof_dev *sdev = dev_get_drvdata(scomp->dev->parent);
+	struct snd_sof_dev *sdev = snd_sof_get_sof_dev(scomp->dev);
```
@ranj063 Have you checked whether we could go even step further and not use "sdev" at all in these platform-specific functions? I see you've cleaned up sdev usage in many places already, so I'm assuming that these are still needed here.

ranj063 replied:
@kv2019i yes I thought about this. The main reason I didnt do this is because of the register reads/writes that the platform-specific code does. I suppose I could add abstracted methods for these as well but do you think its necessary? This is low-level code anyway no?

kv2019i replied:
@ranj063 no, no, just asking whether you've looked into this. And yeah, I now looked better at hda-pcm.c as an example, and yeah, I agreee adding abstractions here does not make sense. The low-level functions have to understand the device as a whole entity (not split into MFD functions, the hw is a combined device after all). So just asking.

RanderWang commented:
> @ranj063 This starts to look really good now. I couldn't spot any obvious problems. I especially appreciate the nice ascii drawing explain how machine driver selection works. That's nice! A rebase is needed and solution to the dma memory alloc issue, but otherwise this seems good to go.

@kv2019i  I also have the same idea.

dbaluta commented:
@ranj063 at audio miniconf Mark & others expressed their concern about the way we are probing machine drivers. Specifically for ARM we should not trigger machine driver probing from SOF core, but to allow this to happen by adding a machine driver node inside the device tree file.

Now, for IMX usecase your patches are just TODOs so I guess would be fine for this patches to be sent for review with ALSA community. Will try to add machine driver probing inside the DTS and annotate your patches again.

Thanks for the awesome work!

plbossart commented:
> @ranj063 at audio miniconf Mark & others expressed their concern about the way we are probing machine drivers. Specifically for ARM we should not trigger machine driver probing from SOF core, but to allow this to happen by adding a machine driver node inside the device tree file.

The current way of probing is forced on us by ACPI/Windows, we don't have a descriptor for the machine driver that we can parse so we invent one.

It's not a deliberate design decision, only a work-around, and there's no desire to push this for other platforms.

dbaluta commented:
> > @ranj063 at audio miniconf Mark & others expressed their concern about the way we are probing machine drivers. Specifically for ARM we should not trigger machine driver probing from SOF core, but to allow this to happen by adding a machine driver node inside the device tree file.
> 
> The current way of probing is forced on us by ACPI/Windows, we don't have a descriptor for the machine driver that we can parse so we invent one.
> 
> It's not a deliberate design decision, only a work-around, and there's no desire to push this for other platforms.

Yes, this is exactly what Intel folks pointed out at the conference and then I really understood the design decision. For IMX we will completely describe the machine drivers in the DT and let kernel core probe them.

Also, I don't think we need the FORCE_CODEC option as we can create a separate DT file to describe the nocodec scenario. With this case we don't even need to recompile the kernel.

plbossart commented:
> Also, I don't think we need the FORCE_CODEC option as we can create a separate DT file to describe the nocodec scenario. With this case we don't even need to recompile the kernel.

Yes, the FORCE_CODEC is only a means to bypass the ACPI checks, it's not needed for the DT platforms.
It can be argued though that for tests you may want to change the kernel and not have to update DT blobs (e.g. for CI).

ranj063 commented:
@kv2019i @RanderWang @plbossart I've addressed all comments and rebased now.
The failure seen on module reload test on BYT nocodec is unrelated to this PR and can be fixed by https://github.com/thesofproject/linux/issues/1424#issuecomment-548963773

stripes416 commented:
@ranj063 good catch & quick work!

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -374,32 +355,13 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
-		dev_name(&plat_data->pdev_mach->dev));
+	/*
+	 * Register audio client.
+	 * This can fail but errors cannot be propagated.
```
the comment is misleading. if there are errors during registration, they will be reported. However errors happening during the platform driver probe will be squelched.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -374,32 +355,13 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+	 */
+	sdev->sof_audio = sof_client_dev_register(sdev, "sof-audio");
+	if (!sdev->sof_audio)
+		dev_warn(sdev->dev, "sof-audio client failed to register\n");
```
you can exit here, but as mentioned above this will not trap errors during the platform driver probe.

ranj063 replied:
@plbossart the idea is that if one client registration fails, it shouldnt affect the others or the top-level device status. If I exit here, that wont be possible isnt it?

plbossart replied:
@ranj063 I prefer it when we fail big and early.
It's the same as trying to continue the probe when i915 is not working. It's a waste of time, there are issues that will pop-up left and right.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -511,25 +472,14 @@ int snd_sof_device_remove(struct device *dev)
 
+	/* Unregister audio client device */
+	if (sdev->sof_audio)
+		sof_client_dev_unregister(sdev->sof_audio);
```
you need to check but for new kernels it's no longer needed to check if the platform_device is ERR or NULL.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -374,32 +355,13 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+	 * This can fail but errors cannot be propagated.
+	 */
+	sdev->sof_audio = sof_client_dev_register(sdev, "sof-audio");
+	if (!sdev->sof_audio)
```
check if platform_register returns ERR_OR_NULL ?

ranj063 replied:
@plbossart I do check for that in sof_client_dev_register()

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,140 @@
+	int ret;
+
+	/* check machine info */
+	ret = sof_machine_check(sdev);
```
use &pdev->dev for the device.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,140 @@
+	plat_data->platform = dev_name(&pdev->dev);
+
+	/* set up platform component driver */
+	snd_sof_new_platform_drv(sdev);
```
don't use sdev in this function.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,140 @@
+	pm_runtime_disable(&pdev->dev);
+
+	if (!IS_ERR_OR_NULL(pdata->pdev_mach))
+		platform_device_unregister(pdata->pdev_mach);
```
check but I don't think we need this check any longer.
And btw if it's needed we need the same for the error on line 95

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,140 @@
+
+static int sof_audio_remove(struct platform_device *pdev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(pdev->dev.parent);
```
not needed

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,140 @@
+
+static struct platform_driver sof_audio_driver = {
+	.driver = {
+		.name = "sof-audio",
```
.pm_ops?

plbossart commented on sound/soc/sof/sof-pci-dev.c:
```diff
@@ -325,7 +325,6 @@ static int sof_pci_probe(struct pci_dev *pci,
 	sof_pdata->machine = mach;
 	sof_pdata->desc = (struct sof_dev_desc *)pci_id->driver_data;
 	sof_pdata->dev = dev;
-	sof_pdata->platform = dev_name(dev);
```
why is this removed?

ranj063 replied:
@plabossart its set in the client device.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -477,6 +477,9 @@ struct snd_sof_dev {
 	bool msi_enabled;
 
+	/* client devices */
+	struct snd_sof_client *sof_audio;
```
should we have num_clients and an array of clients?

ranj063 replied:
@plbossart yes, we should but for this PR I didnt add it just yet.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -793,7 +793,7 @@ static void sof_parse_word_tokens(struct snd_soc_component *scomp,
 				  struct snd_soc_tplg_vendor_array *array)
 {
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_dev *sdev = dev_get_drvdata(scomp->dev->parent);
```
I think we should get rid of sdev in all those topology files first.
Or we should point to a substructure and move it to the client later.
Either, let's not use the parent.

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -264,6 +264,14 @@ static int sof_probe_continue(struct snd_sof_dev *sdev)
+	 */
+	if (!sof_ops(sdev)->runtime_suspend || !sof_ops(sdev)->runtime_resume)
+		pm_runtime_get_noresume(sdev->dev);
+
```
different patch? Why is this here anyways?

ranj063 replied:
yes, probably a different patch. will fix.
But it should have here in the first place. In the current code, this is set after topology loading which is incorrect.

plbossart replied:
move earlier then before the cleanups

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -41,12 +41,15 @@ static int create_page_table(struct snd_soc_component *component,
 {
-	struct snd_sof_dev *sdev = spcm->sdev;
+	struct snd_soc_component *scomp = spcm->scomp;
+	struct snd_sof_dev *sdev = dev_get_drvdata(scomp->dev->parent);
```
remove sdev

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -15,7 +15,36 @@
 #include <sound/sof/topology.h>
-#include "sof-priv.h"
+
+#define DRV_NAME	"sof-audio-component"
```
use a prefix, we use DRV_NAME in all drivers, it's just a mess.

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -18,12 +18,8 @@
 #include <sound/control.h>
```
you should add a comment on who is supposed to use sof-priv.h now? 
Possibly rename it sof-priv-parent.h?

ranj063 replied:
@plbossart sure. Maybe I will rename it to sof-priv-core.h

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -153,7 +152,7 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 	spcm = snd_sof_find_spcm_name(scomp, swidget->widget->sname);
 	if (!spcm) {
-		dev_err(sdev->dev, "error: cannot find PCM for %s\n",
+		dev_err(scomp->dev, "error: cannot find PCM for %s\n",
```
I think we should do this sort of cleanup first, before you even start the multi-client conversation.

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -139,7 +139,6 @@ static int sof_keyword_dapm_event(struct snd_soc_dapm_widget *w,
 {
 	struct snd_sof_widget *swidget = w->dobj.private;
 	struct snd_soc_component *scomp = swidget->scomp;
-	struct snd_sof_dev *sdev = dev_get_drvdata(scomp->dev->parent);
```
this is classic tick-tock effect, we can do better with one set of changes that directly does what we need in the end.

plbossart commented on sound/soc/sof/audio-ops.h:
```diff
@@ -0,0 +1,95 @@
+		       const struct sof_ipc_pcm_params_reply *reply)
+{
+	return sof_audio->audio_ops->ipc_pcm_params(sof_audio->component,
+						    substream, reply);
```
that looks weird to me.
I would have expected this sort of IPC to use the communication channel between client and parent, and then the parent performs the IPC. Directly handling the function is a shortcut that shouldn't exist.

plbossart commented on sound/soc/sof/audio-ops.h:
```diff
@@ -0,0 +1,95 @@
+{
+	if (sof_audio->audio_ops->pcm_pointer)
+		return sof_audio->audio_ops->pcm_pointer(sof_audio->component,
+							 substream);
```
ah, this one is interesting.
We want the client to allocate the buffers, but I would image the parent allocates the DMA stream, so shouldn't the parent get the PCM pointer when it is retrieved from DMA registers. Likewise, if the parent has to look at a shared memory window or trigger an IPC to get the position, it should do so based on a query. I am not sure we want direct access from the client to the hardware, that's asking for trouble and likely will not scale for multiple clients

plbossart commented on sound/soc/sof/intel/byt.c:
```diff
@@ -540,7 +546,7 @@ const struct snd_sof_dsp_ops sof_tng_ops = {
 			SNDRV_PCM_INFO_BATCH,
 };
-EXPORT_SYMBOL(sof_tng_ops);
+EXPORT_SYMBOL(sof_tng_audio_ops);
```
that doesn't look good to me, this is one bridge too far.

plbossart commented on sound/soc/sof/intel/hda-codec.c:
```diff
@@ -80,12 +80,12 @@ EXPORT_SYMBOL(hda_codec_jack_wake_enable);
 /* probe individual codec */
-static int hda_codec_probe(struct snd_sof_dev *sdev, int address)
+static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
+			   struct sof_audio_dev *sof_audio)
```
no, that doesn't look good. the client has nothing to do here.
It's once you've probed the codecs and figured out what platform you are dealing with, then you can create clients. This is doing things the other way around, that's not so good.

plbossart commented on sound/soc/sof/sof-acpi-dev.c:
```diff
@@ -45,6 +45,7 @@ static const struct sof_dev_desc sof_acpi_haswell_desc = {
 	.chip_info = &hsw_chip_info,
 	.default_fw_path = "intel/sof",
 	.default_tplg_path = "intel/sof-tplg",
+	.default_fw_filename = "sof-hsw.ri",
```
so what do we need nocodec_fw_filename for now?

plbossart replied:
ok, so you should have added a comment in earlier patches that the nocodec_filename would be removed after the code changes.

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -48,16 +50,314 @@ static int sof_machine_check(struct snd_sof_dev *sdev)
+
+static int sof_destroy_pipelines(struct device *dev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev->parent);
```
it's still too confusing to rely on sdev, Let's remove this dependency first.

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -0,0 +1,16 @@
+
+	return client->client_data;
+}
+EXPORT_SYMBOL(sof_get_client_data);
```
do really need this? A static inline would just work well here.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -306,6 +306,18 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
+			  void *msg_data, size_t msg_bytes, void *reply_data,
+			  size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev->parent);
```
that's the only place where you'd want sdev :-)

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -207,6 +210,7 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 	struct sof_intel_hda_stream *hda_stream;
 	struct hda_pipe_params p_params = {0};
 	struct hdac_ext_link *link;
+	struct snd_soc_component *scomp = snd_soc_rtdcom_lookup(rtd, DRV_NAME);
```
what happens if you have two audio clients?
the use of DRV_NAME was fine when there was no ambiguity, now it's not enough.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -513,11 +513,11 @@ int snd_sof_ipc_stream_posn(struct snd_sof_dev *sdev,
-	err = sof_ipc_tx_message(sdev->ipc,
-				 stream.hdr.cmd, &stream, sizeof(stream), &posn,
-				 sizeof(*posn));
+	err = sof_client_tx_message(scomp->dev,
```
you could add these changes before adding the client/parent split, that would help validate before the big bang.

plbossart commented on sound/soc/sof/sof-client.h:
```diff
@@ -14,6 +14,10 @@
 
+	/* IPC RX callback */
+	void (*sof_client_rx_cb)(struct snd_sof_client *client,
+				 u32 msg_cmd);
```
can you not get more than a u32 back?
what is msg_cmd used for?

plbossart commented on sound/soc/sof/sof-priv.h:
```diff
@@ -488,6 +488,12 @@ struct snd_sof_dev {
 };
 
+struct ipc_rx_client {
+	u32 ipc_cmd;
```
what does ipc_cmd represent? A client ID? something else?

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -338,6 +338,114 @@ static const char *fixup_tplg_name(struct snd_sof_dev *sdev,
 }
 
+int hda_machine_driver_select(struct snd_sof_dev *sdev,
+			      struct sof_audio_dev *sof_audio)
```
again I don't like mixing parent and child. this is not the right way to do this.

ranj063 commented:
> . I don't like the audio_ops. We should use the client IPC channel and not directly handle hardware resources.

@plbossart I didnt quite catch what you meant here. For example, lets take the pm_trigger() op, there is no IPC in hda_dsp_stream_trigger(). All it does is a few register reads/writes to start/stop the DMA.
How can I use the IPC channel for this?

ranj063 commented:
> I don't think you can use DRV_NAME to identify components, if you have two audio clients (or more) you will pick the first (and most likely wrong) component from a list.

@plbossart I dont think this is true. Every audio client creates its own card and the components are searched within the card. So even if the names are the same, the component itself will be different.
I have already verified this as I have the multiple client version working with my PR.

ranj063 commented:
> I see too many functions that deal with both sof-dev and sof-audio, this is not good in terms of functionality split. we should have clearly defined interfaces between the two parts, handling pointers makes the split less intuitive and is a recipe for disaster in general.

The only code that deals with both sof-dev and sof-audio-dev is the low-level platform-specific stuff. I have consciously removed sof-dev from all client code by the end of the series. Agreed its not easy to tell from just reviewing the first few commits in the series but the if you checkout the PR and look into the code, it will make it much easier for you to see this more clearly.

Anyway, like I mentioned before it was quite hard for me to isolate the sof-dev parts from the client code from the very first commit. I could give it another try and see if I can do it differently.

plbossart commented:
> > . I don't like the audio_ops. We should use the client IPC channel and not directly handle hardware resources.
> 
> @plbossart I didnt quite catch what you meant here. For example, lets take the pm_trigger() op, there is no IPC in hda_dsp_stream_trigger(). All it does is a few register reads/writes to start/stop the DMA.
> How can I use the IPC channel for this?

The streams should be allocated and managed at the PCI level, the client should not muck with them.

ranj063 commented:
> The streams should be allocated and managed at the PCI level, the client should not muck with them.

@plbossart Agree but that's what this PR already does. The client does not touch the streams.
But I separated out the audio_ops to the client because a pcm_trigger or a pcm_params is audio specific. This way I can isolate the core ops from the client as well.
Are you saying that the there is no need to partition the audio-specific ops from the core ops?

plbossart commented:
> > The streams should be allocated and managed at the PCI level, the client should not muck with them.
> 
> @plbossart Agree but that's what this PR already does. The client does not touch the streams.
> But I separated out the audio_ops to the client because a pcm_trigger or a pcm_params is audio specific. This way I can isolate the core ops from the client as well.
> Are you saying that the there is no need to partition the audio-specific ops from the core ops?

I don't see why you would need to expose to the client audio specific ops. expose an API and leave the details to the core.
We must be talking past each other on this one.

ranj063 commented:
> I don't see why you would need to expose to the client audio specific ops. expose an API and leave the details to the core.
> We must be talking past each other on this one.

@plbossart OK makes sense. I will move to the client API's.

ranj063 commented:
@plbossart I have to pretty much reverse the order of the commits in this PR for it to make sense in terms of reviewing and not having to make changes and overwrite them. Let me work on it next and also address your other comments while I work on this.

ranj063 commented:
@plbossart @dbaluta @kv2019i I've redone this PR starting with the clean-up first and then adding the client towards the end. There's still some minor issues that I need to address based one earlier comments.

plbossart commented:
> @plbossart @dbaluta @kv2019i I've redone this PR starting with the clean-up first and then adding the client towards the end. There's still some minor issues that I need to address based one earlier comments.

Did you mean we can review now, or you'll notify us when the changes are ready?

ranj063 commented:
> Did you mean we can review now, or you'll notify us when the changes are ready?

@plbossart yes, ready for review.

ranj063 commented:
SOFCI TEST

dbaluta commented on sound/soc/sof/sof-audio.c:
```diff
@@ -556,27 +635,22 @@ static int sof_audio_probe(struct platform_device *pdev)
+	/* select machine driver */
+	ret = sof_audio_select_machine(pdev, desc);
+	if (ret < 0)
 		return ret;
```
@ranj063 because machine driver for IMX will be probed separately by placing a description in DT I think all the functions related to machine drivers in sof-audio.c should be no-ops for IMX.

Sorry, for misleading you. Will try to send some patches on top of this PR.

Not sure we even need the sof_fw_mach structure now.

There are still some things I need to figure out:
* if machine driver probing is decoupled from SOF core how do we set .ignore machine field?
* platform_device_register_data that creates machine device for ACPI / PCI should be abstracted as a per device ops, with no-op for IMX

ranj063 replied:
@dbaluta no worries. this PR is still under review. So please let me know when you have more info.

dbaluta replied:
@ranj063 I have a wild proposal for you. What do you think if I base my patches on current sof-dev branch without this PR, then you rebase your patches on mine.

Thus it would be much easier for me to backport support for IMX machine driver in 5.4

This is just a thought.

ranj063 replied:
@dbaluta I think thats better. I will undo the common sof_fw_mach changes in this series and wait for your PR

dbaluta replied:
great, thanks!

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,349 @@
+	 * So, set the flag to indicate this for those streams that
+	 * have been suspended.
+	 */
+	list_for_each_entry(spcm, &sdev->pcm_list, list) {
```
you've added a pcm_list in snd_audio_dev, so is this really needed to use sdev here?

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -213,34 +28,6 @@ static int sof_send_pm_ctx_ipc(struct snd_sof_dev *sdev, int cmd)
-}
-
 #if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_ENABLE_DEBUGFS_CACHE)
 static void sof_cache_debugfs(struct snd_sof_dev *sdev)
```
wondering if this should be split, aren't there debugfs parts that the client should own?

plbossart commented on sound/soc/sof/pm.c:
```diff
@@ -346,7 +133,7 @@ static int sof_suspend(struct device *dev, bool runtime_suspend)
 	/* set restore_stream for all streams during system suspend */
 	if (!runtime_suspend) {
-		ret = sof_set_hw_params_upon_resume(sdev);
+		ret = sof_set_hw_params_upon_resume(sdev->dev);
```
you've moved this to sof-audio so not sure why we use sdev here.

plbossart commented on sound/soc/sof/sof-audio.h:
```diff
@@ -0,0 +1,33 @@
+
+/* SOF audio device */
+struct sof_audio_dev {
+	/*
```
should there be a struct device *dev here?

that would be very handy for code changes and remove the need for sdev.

dbaluta replied:
@ranj063 any comments on @plbossart's remark?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,349 @@
+
+static int sof_restore_kcontrols(struct device *dev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
```
sof_audio_dev?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -0,0 +1,349 @@
+
+int sof_set_hw_params_upon_resume(struct device *dev)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
```
sof_audio_dev?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -307,6 +307,18 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
 EXPORT_SYMBOL(sof_ipc_tx_message);
 
+/* send IPC message from client to DSP */
+int sof_client_tx_message(struct device *dev, u32 header,
```
what does 'dev' represent here? the client or the parent?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -307,6 +307,18 @@ int sof_ipc_tx_message(struct snd_sof_ipc *ipc, u32 header,
+			  void *msg_data, size_t msg_bytes, void *reply_data,
+			  size_t reply_bytes)
+{
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
```
probably need a different macro to get the parent?

plbossart replied:
edit: in next patches you use snd_sof_get_sof_dev, this should be added first and this patch shouldn't use dev_get_drvdata()

plbossart commented on sound/soc/sof/core.c:
```diff
@@ -367,11 +370,26 @@ int snd_sof_device_probe(struct device *dev, struct snd_sof_pdata *plat_data)
+
+	/* create SOF audio client data */
+	sof_audio = devm_kzalloc(sdev->dev, sizeof(*sof_audio), GFP_KERNEL);
+	if (!sof_audio)
```
if the probe fails, don't you need to free sdev->sof_audio?
The .remove will not be called so the devm stuff will not reclaim the memory?

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,14 +154,16 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 	int ret = 0;
 
-	list_for_each_entry(sof_dai, &hda_stream->sdev->dai_list, list) {
+	list_for_each_entry(sof_dai, &sof_audio->dai_list, list) {
```
can we move this to an earlier patch, thsi doesn't seem related to the partition of sdev, it's all the component stuff.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -184,11 +187,11 @@ static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
 				dev_err(hda_stream->sdev->dev,
 					"error: failed to set dai config for %s\n",
 					sof_dai->name);
+
```
newlines

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -224,7 +228,7 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 
 	/* update the DSP with the new tag */
-	ret = hda_link_config_ipc(hda_stream, dai->name, stream_tag - 1,
+	ret = hda_link_config_ipc(hda_stream, scomp, dai->name, stream_tag - 1,
```
this can also be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -273,7 +276,7 @@ static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
 		return 0;
 
-	dev_dbg(sdev->dev, "hda: prepare stream dir %d\n", substream->stream);
+	dev_dbg(scomp->dev, "hda: prepare stream dir %d\n", substream->stream);
```
and this.

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -354,19 +357,19 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
-	struct snd_soc_pcm_runtime *rtd;
+	struct snd_soc_pcm_runtime *rtd = snd_pcm_substream_chip(substream);
 	struct hdac_ext_stream *link_dev;
+	struct snd_soc_component *scomp = dai->component;
```
this as well

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -354,19 +357,19 @@ static int hda_link_hw_free(struct snd_pcm_substream *substream,
-	ret = hda_link_config_ipc(hda_stream, dai->name, DMA_CHAN_INVALID,
-				  substream->stream);
+	ret = hda_link_config_ipc(hda_stream, scomp, dai->name,
+				  DMA_CHAN_INVALID, substream->stream);
```
and this

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -122,7 +122,7 @@ static int sof_pcm_hw_params(struct snd_soc_component *component,
 	ret = snd_pcm_lib_malloc_pages(substream, params_buffer_bytes(params));
 	if (ret < 0) {
-		dev_err(component->dev, "error: could not allocate %d bytes for PCM %d\n",
+		dev_err(component->dev,	"error: could not allocate %d bytes for PCM %d\n",
```
spurious diff?

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -627,7 +627,8 @@ static int sof_pcm_new(struct snd_soc_component *component,
 	snd_pcm_lib_preallocate_pages(pcm->streams[stream].substream,
-				      SNDRV_DMA_TYPE_DEV_SG, sdev->dev,
+				      SNDRV_DMA_TYPE_DEV_SG,
+				      component->dev,
```
all this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2366,7 +2368,7 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
 
 	/* allocate playback page table buffer */
-	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, sdev->dev,
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, scomp->dev,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2396,7 +2398,7 @@ static int sof_dai_load(struct snd_soc_component *scomp, int index,
 
 	/* allocate capture page table buffer */
-	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, sdev->dev,
+	ret = snd_dma_alloc_pages(SNDRV_DMA_TYPE_DEV, scomp->dev,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2503,7 +2506,7 @@ static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
 	 */
 	if (!found) {
-		dev_warn(sdev->dev, "warning: failed to find dai for dai link %s",
+		dev_warn(scomp->dev, "warning: failed to find dai for dai link %s",
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2474,14 +2476,15 @@ static void sof_dai_set_format(struct snd_soc_tplg_hw_config *hw_config,
 
 /* set config for all DAI's with name matching the link name */
-static int sof_set_dai_config(struct snd_sof_dev *sdev, u32 size,
+static int sof_set_dai_config(struct snd_soc_component *scomp, u32 size,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2579,7 +2582,7 @@ static int sof_link_ssp_load(struct snd_soc_component *scomp, int index,
 
 	/* set config for all DAI's with name matching the link name */
-	ret = sof_set_dai_config(sdev, size, link, config);
+	ret = sof_set_dai_config(scomp, size, link, config);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2657,7 +2660,7 @@ static int sof_link_esai_load(struct snd_soc_component *scomp, int index,
 
 	/* set config for all DAI's with name matching the link name */
-	ret = sof_set_dai_config(sdev, size, link, config);
+	ret = sof_set_dai_config(scomp, size, link, config);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2797,17 +2801,19 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
  * and send config to FW. In FW, each dai sets config by dai_index
  */
-static int sof_link_hda_process(struct snd_sof_dev *sdev,
+static int sof_link_hda_process(struct snd_soc_component *scomp,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2797,17 +2801,19 @@ static int sof_link_dmic_load(struct snd_soc_component *scomp, int index,
 				struct snd_soc_dai_link *link,
 				struct sof_ipc_dai_config *config)
 {
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2885,7 +2890,7 @@ static int sof_link_hda_load(struct snd_soc_component *scomp, int index,
 	}
 
-	ret = sof_link_hda_process(sdev, link, config);
+	ret = sof_link_hda_process(scomp, link, config);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -2919,7 +2924,7 @@ static int sof_link_alh_load(struct snd_soc_component *scomp, int index,
 
 	/* set config for all DAI's with name matching the link name */
-	ret = sof_set_dai_config(sdev, size, link, config);
+	ret = sof_set_dai_config(scomp, size, link, config);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -3050,15 +3055,15 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 }
 
-static int sof_link_hda_unload(struct snd_sof_dev *sdev,
+static int sof_link_hda_unload(struct snd_soc_component *scomp,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -3050,15 +3055,15 @@ static int sof_link_load(struct snd_soc_component *scomp, int index,
 	dai = snd_soc_find_dai(link->cpus);
 	if (!dai) {
-		dev_err(sdev->dev, "error: failed to find dai %s in %s",
+		dev_err(scomp->dev, "error: failed to find dai %s in %s",
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/topology.c:
```diff
@@ -3100,7 +3105,7 @@ static int sof_link_unload(struct snd_soc_component *scomp,
 		break;
 	case SOF_DAI_INTEL_HDA:
-		ret = sof_link_hda_unload(sdev, link);
+		ret = sof_link_hda_unload(scomp, link);
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -153,14 +154,16 @@ static int hda_link_dma_params(struct hdac_ext_stream *stream,
 
 /* Send DAI_CONFIG IPC to the DAI that matches the dai_name and direction */
 static int hda_link_config_ipc(struct sof_intel_hda_stream *hda_stream,
+			       struct snd_soc_component *scomp,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -208,6 +211,7 @@ static int hda_link_hw_params(struct snd_pcm_substream *substream,
 	struct sof_intel_hda_stream *hda_stream;
 	struct hda_pipe_params p_params = {0};
 	struct hdac_ext_link *link;
+	struct snd_soc_component *scomp = dai->component;
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -263,8 +267,7 @@ static int hda_link_pcm_prepare(struct snd_pcm_substream *substream,
 	struct sof_intel_hda_stream *hda_stream;
-	struct snd_sof_dev *sdev =
-				snd_soc_component_get_drvdata(dai->component);
+	struct snd_soc_component *scomp = dai->component;
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -285,7 +288,8 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 	struct sof_intel_hda_stream *hda_stream;
-	struct snd_soc_pcm_runtime *rtd;
+	struct snd_soc_pcm_runtime *rtd = snd_pcm_substream_chip(substream);
+	struct snd_soc_component *scomp = dai->component;
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/intel/hda-dai.c:
```diff
@@ -324,7 +327,7 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 		 * hw_params is set up again after resume.
 		 */
-		ret = hda_link_config_ipc(hda_stream, dai->name,
+		ret = hda_link_config_ipc(hda_stream, scomp, dai->name,
```
this can be moved to an earlier patch

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -16,7 +16,6 @@
 #include "sof-priv.h"
```
commit message: Preparation

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -503,8 +498,8 @@ static int sof_pcm_open(struct snd_soc_component *component,
-	/* set runtime config */
-	runtime->hw.info = ops->hw_info; /* platform-specific */
+	/* set platform-specific runtime config */
+	runtime->hw.info = snd_sof_get_hw_info(component->dev);
```
snd_sof_client_get_hw_info() ?

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -319,6 +319,16 @@ int sof_client_tx_message(struct device *dev, u32 header,
 
+/* register client to receive IPC */
+void snd_sof_ipc_rx_register(struct snd_sof_client *client,
+			     struct device *dev)
```
it'd be nicer if we have a dev field in snd_sof_client, that way the device information is hidden, and you'd use an indirection client->dev
'dev' alone requires the reviewer to look at the call tree to figure out what the arguments were, it's not that self-explanatory.

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -52,6 +52,9 @@ config SND_SOC_SOF_DEVELOPER_SUPPORT
 	  Say Y if you are involved in SOF development and need this option
```
I don't know how you format your commit messages but you can use more words per line...

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -52,6 +52,9 @@ config SND_SOC_SOF_DEVELOPER_SUPPORT
 	  If not, select N
 
+config SND_SOC_SOF_AUDIO
+	tristate
```
add some help to describe the option

plbossart commented on sound/soc/sof/Kconfig:
```diff
@@ -179,6 +182,7 @@ config SND_SOC_SOF
 	tristate
 	select SND_SOC_TOPOLOGY
 	select SND_SOC_SOF_NOCODEC if SND_SOC_SOF_NOCODEC_SUPPORT
+	select SND_SOC_SOF_AUDIO
```
Ah that's an interesting one.
Do we *always* require a separate platform device, or do we allow in some cases the hierarchy to remain with a single device?
With all the work you've done, I am not sure if we actually require the platform device in all cases.

It's the same e.g. as workqueues for probe, it's required for HDAudio but not for all platforms.

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -517,7 +447,7 @@ int snd_sof_ipc_stream_posn(struct snd_soc_component *scomp,
 			    struct sof_ipc_stream_posn *posn)
 {
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_dev *sdev = snd_sof_get_sof_dev(scomp->dev);
```
move to earlier patch

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -530,6 +460,7 @@ int snd_sof_ipc_stream_posn(struct snd_soc_component *scomp,
 	err = sof_ipc_tx_message(sdev->ipc,
 				 stream.hdr.cmd, &stream, sizeof(stream), &posn,
 				 sizeof(*posn));
+
```
spurious newline

plbossart commented on sound/soc/sof/ipc.c:
```diff
@@ -654,7 +582,7 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_control *scontrol,
 	struct snd_soc_component *scomp = scontrol->scomp;
 	struct sof_ipc_ctrl_data *cdata = scontrol->control_data;
-	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_dev *sdev = snd_sof_get_sof_dev(scomp->dev);
```
move to earlier patch

plbossart commented on sound/soc/sof/intel/hda-pcm.c:
```diff
@@ -149,12 +149,14 @@ snd_pcm_uframes_t hda_dsp_pcm_pointer(struct snd_sof_dev *sdev,
 	snd_pcm_uframes_t pos;
 
+	scomp = snd_soc_rtdcom_lookup(rtd, SOF_AUDIO_PCM_DRV_NAME);
+
```
move to earlier patch?

plbossart commented on sound/soc/sof/sof-audio.c:
```diff
@@ -347,3 +467,180 @@ struct snd_sof_dai *snd_sof_find_dai(struct snd_soc_component *scomp,
+	SET_RUNTIME_PM_OPS(sof_audio_runtime_suspend, sof_audio_resume, NULL)
+};
+
+static int sof_audio_probe(struct platform_device *pdev)
```
we could use a wrapper so that we can reuse the same code as before?

e.g. 
sof_audio_probe_pdev(struct platform_device *pdev) 
{
    return sof_audio_probe(&pdev->dev)
}

that would help test without necessarily create this platform driver

plbossart commented on include/sound/soc-of.h:
```diff
@@ -0,0 +1,22 @@
+/* SPDX-License-Identifier: GPL-2.0
```
you see to use fewer and fewer words per line in commit messages :-)

plbossart commented on include/sound/soc-of.h:
```diff
@@ -0,0 +1,22 @@
+ */
+struct snd_soc_of_mach {
+	const char *drv_name;
+	struct device_node *of;
```
we should really drop the soc_acpi stuff and call it soc_firmware or something.
I don't mind if we do this first, but I think we should avoid using two structures when we can use one...

plbossart commented on sound/soc/sof/sof-client.c:
```diff
@@ -225,3 +225,12 @@ int snd_sof_client_ipc_pcm_params(struct device *dev,
+
+	return snd_sof_machine_driver_select(sdev);
+}
+EXPORT_SYMBOL(snd_sof_client_machine_driver_select);
```
this sort of wrappers can be a static inline

plbossart commented on include/sound/sof.h:
```diff
@@ -92,9 +92,7 @@ static inline int sof_mach_get_mach_size(const struct snd_soc_sof_mach *mach)
  */
```
overall comment for this patch: there are too many changes here, it's a bit too hard to review. We should try and split in steps if possible

plbossart commented on sound/soc/sof/pcm.c:
```diff
@@ -574,6 +574,7 @@ static int sof_pcm_close(struct snd_soc_component *component,
 static int sof_pcm_new(struct snd_soc_component *component,
```
can we add in the commit message a reminder of how memory allocation works if we use the parent to allocate the memory? I recall the ALSA core willl reclaim the memory when the card is closed, but how does it work if this is allocated by a different device?

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -414,11 +415,24 @@ struct snd_soc_sof_mach *hda_machine_driver_select(struct snd_sof_dev *sdev)
+
+void hda_set_mach_params(struct snd_sof_dev *sdev,
+			 struct snd_soc_acpi_mach_params *mach_params)
+{}
```
good idea to be able to set the machine though, this can be useful for quirks.

plbossart commented on sound/soc/sof/intel/hda.c:
```diff
@@ -414,11 +415,24 @@ struct snd_soc_sof_mach *hda_machine_driver_select(struct snd_sof_dev *sdev)
 	return NULL;
 }
+
+void hda_set_mach_params(struct snd_sof_dev *sdev,
```
the use of sdev is not quite right here if we want the client to make the machine driver selection?

ranj063 commented:
@plbossart thanks for the review. I finally rootcaused the module loading/unloading problem too. I will work on the next version today/tomorrow.

ranj063 commented:
SOFCI TEST

dbaluta commented on sound/soc/sof/core.c:
```diff
@@ -220,15 +220,15 @@ EXPORT_SYMBOL(snd_sof_get_status);
  * into compressed page table.
```
@ranj063 if there is no trouble, I would really like you to state in the commit message WHY is this change needed.

dbaluta commented on sound/soc/sof/sof-client.h:
```diff
@@ -0,0 +1,17 @@
+/* SPDX-License-Identifier: (GPL-2.0 OR BSD-3-Clause) */
+/*
```
@ranj063 this patch doesn't contain the register/unregister method? Is it intended? Then you will need to update the commit message.

dbaluta commented on sound/soc/sof/sof-priv.h:
```diff
@@ -585,8 +585,6 @@ int snd_sof_ipc_set_get_comp_data(struct snd_sof_ipc *ipc,
  * There is no snd_sof_free_topology since topology components will
```
@ranj063 some of the patches are independent of this PR and I think we can already merge them to decrease the number of patches in the PR and to avoid re-reviewing them all the time.

plbossart replied:
yes, I'd really like a 'prepare multi-client' PR so that we can decrease the volume of code. this is what we did for S0ix and it was much easier. Anything about 10 commits is just awful on GitHub, and even in emails it's borderline manageable.

ranj063 commented:
closing this. will need to be replaced with a new one
